rust/src/utils.rs

```diff
@@ -44,10 +44,36 @@ fn download_url_to_tmpfile(url: &str) -> Fallible<fs::File> {
     Ok(tmpf)
 }
 
-/// Open file and provide context containing filename on failures.
+/// Open file for reading and provide context containing filename on failures.
 pub fn open_file<P: AsRef<Path>>(filename: P) -> Fallible<fs::File> {
     return Ok(fs::File::open(filename.as_ref())
-        .with_context(|e| format!("Can't open file {:?}: {}", filename.as_ref().display(), e))?);
+        .with_context(|e| format!("Can't open file {:?} for reading: {}", filename.as_ref().display(), e))?);
+}
+
+/// Open file for writing and provide context containing filename on failures.
+pub fn create_file<P: AsRef<Path>>(filename: P) -> Fallible<fs::File> {
+    return Ok(fs::File::create(filename.as_ref())
+        .with_context(|e| format!("Can't open file {:?} for writing: {}", filename.as_ref().display(), e))?);
+}
+
+/// Open file for writing, passes a Writer to a closure, and closes the file, with O_TMPFILE
+/// semantics.
+pub fn write_file<P, F>(
+    filename: P,
+    f: F
+) -> Fallible<()>
+where
+    P: AsRef<Path>,
+    F: Fn(&mut io::BufWriter<&mut fs::File>) -> Fallible<()>,
```

Comment:
I couldn't figure out the correct combination to express a closure that takes a concrete type that implements `Write` here. I tried the obvious thing:

```rust
pub fn write_file<P, F, W>(
    filename: P,
    f: F
) -> Fallible<()>
where
    P: AsRef<Path>,
    W: Write,
    F: Fn(&mut W) -> Fallible<()>,
{
```

but it still gives a type mismatch error:

```error[E0308]: mismatched types
  --> src/utils.rs:87:11
   |
87 |         f(&mut w)?;
   |           ^^^^^^ expected type parameter, found struct `std::io::BufWriter`
   |
   = note: expected type `&mut W`
              found type `&mut std::io::BufWriter<&mut std::fs::File>`
```

I'm sure there's a way to do this without resorting to trait objects, but I didn't want to spend much more time on it right now.